Implement "Recently Used Boards" Menu

NPoole opened the pull request:
This PR adds a "Recently Used Boards" section to the "Boards" submenu. The effect of this change is to add a list of 5 menu items to the "boards" submenu which represents the 5 most recently selected boards. When clicked, these menu items select the board in question. An empty "Recently Used Boards" section will never appear and will only render after a board has been selected by the user, populating the `recent.boards` preferences key.

**Implementation**
- During `rebuildBoardsMenu()`, check to see whether preferences has `recent.boards`. If so, insert separators and a menu label to denote the "Recently Used Boards" list and store the index of this section for the insertion of menu items later.
- During `createBoardMenusAndCustomMenus(...)` create a collection of 'clone' menu items whose action is to click the original menu items. Reference them by boardId. 
- During `onBoardOrPortChange()` update the list of boardIds in `recent.boards`. This list is treated as a stack with new boards popping off the oldest member of the list. Then call `rebuildRecentBoardsList()` to remove previous menu items and insert new menu items using the collection generated during `createBoardMenusAndCustomMenus(...)`. 

**Notes**
- Broadens the scope of `JMenu boardMenu`
- Adds a small collection to preferences.txt

facchinm commented:
Hi @NPoole ,
were you aware of this https://github.com/arduino/Arduino/pull/8607 ? :slightly_smiling_face: 
Anyway, your patch is by far better since it implements persistence! 
The only thing I suggest is adding an accelerator on the menu entries (https://github.com/arduino/Arduino/pull/8607/files#diff-b6fda0bf6be0a53c0741a9b02ac16f37R1400) to speedup the switch between two boards

matthijskooijman reviewed (commented):
Yeah, I really want to get this feature merged. There was indeed already a PR, but it was not quite ideal and had to be refactored after my board submenu change from a while back, so I'm all for going ahead with this PR. This PR also seems to be signficantly simpler, possibly because it cleverly reuses the existing board menu items and actions.

Overall, it looks good to me. I left a few inline suggestions for improvement.

One significant question I have, is how this should handle board menus. Currently, only the board id itself is saved, AFAICS. However, when you select e.g. a Nano, then select the 168 version instead of the 328 version, then switch to another board and then switch back to Nano in the recent boards list, you might expect that it remembers that you used the 168 version.

Doing this would also make this recent boards list a lot more powerful, especially for boards with a lot of options, such as the ESP series, or the STM32 core that uses the "board" selection to select a family and selects the actual board using a board option.

To implement this, I guess whenever options are selected, the full fqbn including options must be put into the recent list, with some extra care to only compare the basic board id part of the fqbn when removing a board from the recent boards list.

This does mean that when you use a single board with different options, it will only end up in the recent list twice, but that seems acceptable to me. If you would want to include it twice, you would have a risk of including even more entries for the same board (if you need to set more than one option, can be mitigated by only actualy setting the options when you verify or upload, but that will lead to confusion). It is also quite non-trivial to come up with a proper label for the recent boards menu that allows distinguishing the same board with different options, so I think just including each board once (but *do* remember the last-used options) is the better option here.

Note that this same expectation could be had from the regular boards menu, but there you could easily expect that selecting the same board as earlier gets you the same default settings again. When selecting a board from a "recently used boards" list, I would be more inclined to expect to also select the settings for the board that I recently used.

NPoole commented on app/src/processing/app/Base.java:
```diff
@@ -1360,6 +1367,26 @@ public void onBoardOrPortChange() {
+      if (!recentBoard.equals(currentBoard)) {
+        newRecentBoardIds.add(recentBoard);
+      }
+    }
```
I'll look into that. I'm actually not that familiar with Java, this was my first crack at it. I suspect there are a lot of optimizations to do with collections since I mostly treated them like arrays.

NPoole replied:
I can't figure out the optimization for this without breaking it.

matthijskooijman replied:
I implemented this in my branch: https://github.com/matthijskooijman/Arduino/commit/818c3fe3a497ea75737ff83c0631b943fa0f8413

NPoole commented on app/src/processing/app/Base.java:
```diff
@@ -1458,6 +1486,16 @@ public void actionPerformed(ActionEvent actionevent) {
     boardsCustomMenus.add(boardMenu);
 
+    // Insert recently used boards menu and remember index for insertion later
+    if (PreferencesData.has("recent.boards")) {
```
This is a good point. In my head, it seemed ugly to leave a menu header without any entries underneath but you make a good point about there always being a board selected. This could become a check for `recent.num_boards == 0`

NPoole commented on app/src/processing/app/Base.java:
```diff
@@ -1555,6 +1593,23 @@ private String getPlatformUniqueId(TargetPlatform platform) {
+  // menu and repopulate with updated items
+  private void rebuildRecentBoardsList() throws Exception {
+    Collection<String> recentBoardIds = new LinkedList<>();
+    recentBoardIds = PreferencesData.getCollection("recent.boards");
```
Entirely possible

NPoole replied:
Yup, unnecessary.

NPoole commented on app/src/processing/app/Base.java:
```diff
@@ -1585,6 +1640,21 @@ public void actionPerformed(ActionEvent actionevent) {
+    }; 
+
+    // create a menu item for the "recent boards" menu
+    JMenuItem itemClone = new JMenuItem(actionClone);
```
This was my original intention but it gets a little complicated. It requires creating a `ButtonGroup` every time we update the list and adding all of the new buttons to it, which isn't wild, but it also requires that menu to "know" when a board that isn't on the recent list is selected so that it can manually unselect all boards in the recent list. I couldn't come up with an elegant way to do this on first blush but if people think it's worthwhile, I can give it another go. I just had nightmares of the radio buttons getting out of sync between the recent list and the all boards list.

matthijskooijman replied:
Isn't this something that `rebuildRecentBoardsList` could do? Just select the first item and deselect all others? Or more generically, select any item that matches the current item and deselect others (which should be effectively the same)?

NPoole commented on app/src/processing/app/Base.java:
```diff
@@ -1585,6 +1640,21 @@ public void actionPerformed(ActionEvent actionevent) {
+    JMenuItem itemClone = new JMenuItem(actionClone);
+
+    // populate list of menuitem copies
+    recentBoardItems.put(boardId, itemClone);
```
The problem I have is that during `createBoardMenusAndCustomMenus()` I don't know what is going to be on the recent boards list, and the list is updated more often than `createBoardMenusAndCustomMenus()` is called. Whether I save the item, a copy of the item, or just the action, I need to save one for every board that's installed. There are only two solutions I can imagine:

1) Add a function that drills down through all installed packages and calls `createBoardMenusAndCustomMenus()`. This new function would probably be called in `rebuildRecentBoardsList()`. I think this would significantly slow down switching boards or ports so probably not a good idea.

2) Instead of cloning and saving the menu items, retrieve them from the menu during `rebuildRecentBoardsList()`. Presumably, they live somewhere? I spent a long time trying to figure out how to do this... never did.

matthijskooijman replied:
> Whether I save the item, a copy of the item, or just the action, I need to save one for every board that's installed.

Yes, I was suggesting you still have a map that contains an item for each board in the entire menu, but you wouldn't have to create the clone item and action for each, just for the ones you're actually going to use.

You could probably also find them without the map, but just using the map is probably a lot faster then trying to find them from the menus.

matthijskooijman replied:
I implemented this locally, cleans up things nicely. However, some other small improvements made broke things, so I haven't pushed this anywhere yet, no time to fix that now, so I'll come back to this. Also the custom board menu stuff might also need some further changes here.

matthijskooijman replied:
I implemented this in my branch https://github.com/matthijskooijman/Arduino/commit/a9631da193a4d0325020d753c310a29d2ccfc421

Note that when stuff is refactored by saving `fqbn`'s, as I suggest in #10887, this might end up different again, though.

NPoole commented:
Sorry for letting this languish for so long, I had other work to attend to. 

I've addressed most of your comments, I think, in the latest commit. 

As for logging different permutations of board custom menus, that feels like feature creep. I agree that it feels like expected behavior but at the same time, I can't think of a clean way to represent the different permutations in the menu. Some boards have 4 or more custom menus so while it's not totally outlandish to image an entry like:
"Arduino Pro or Pro Mini [ATmega328P (5V 16MHz)]"
You would also expect entries in excess of:
"SparkFun Artemis Module[SparkFun Variable Bootloader(Recommended),921600]"

Otherwise, you get identical entries that represent different configurations but with no indication of which configuration they represent. Do you see what I mean? At the point it becomes useful, it also becomes cumbersome. 

The original purpose of this PR is to solve the problem of diving too deep into the board menus, and I think it does that. The board custom menus don't require the user to open nested drop-downs and, so far, don't contain large numbers of options to scroll through. Respectfully.

matthijskooijman commented on app/src/processing/app/Base.java:
```diff
@@ -1360,6 +1367,39 @@ public void onBoardOrPortChange() {
+    }
+    newRecentBoardIds.add(0, currentBoard);
+
+    int numBoards = 0;
+
+    if (PreferencesData.has("recent.num_boards")) {
+      numBoards = PreferencesData.getInteger("recent.num_boards");
+    }
```
`getInteger` can take a default argument, so this can be:

```suggestion
    int numBoards = PreferencesData.getInteger("recent.num_boards", 0);
```

But, given what I say about the default `preferences.txt` in another comment, I think the default can even be left out:

```suggestion
    int numBoards = PreferencesData.getInteger("recent.num_boards");
```

matthijskooijman replied:
I did this in my branch, as part of some other cleanup.

matthijskooijman commented on app/src/processing/app/Base.java:
```diff
@@ -1376,16 +1376,29 @@ public void onBoardOrPortChange() {
+    // If recent.num_boards is 0, interpret this as the feature
+    // being turned off. There's no need to rebuild the menu
+    // because it will be hidden
+    if (numBoards > 0) {
```
I believe this if could be broader: If `numBoards == 0`, then there's no point to even set up, add to and prune the list at all. It might still be needed to create an empty list to keep other code happy, but maybe not even.

matthijskooijman replied:
In my branch, I ended up just removing this `if` instead, since the code should produce the right result in that case anyway (it is slightly slower, but having this disabled is a corner case that should not really be optimized for).

matthijskooijman commented on app/src/processing/app/Base.java:
```diff
@@ -1458,12 +1499,30 @@ public void actionPerformed(ActionEvent actionevent) {
     boardsCustomMenus.add(boardMenu);
 
+    // Insert recently used boards menu and remember index for insertion later
+    // Check if the field exists, in case preferences got lost
+    if (!PreferencesData.has("recent.num_boards")){
+        // (default to 5)
+          PreferencesData.setInteger("recent.num_boards", 5);
+    }
```
I think that setting default values for preferences like this is not normally done, instead it should be added to `build/shared/lib/preferences.txt`, which is always loaded *before* the user's `preferences.txt` IIRC, so that sets the default value for when the user's file does not contain the preference (which is then saved to the user's `preferences.txt` on quit).

So this could just be removed.

```suggestion
```

matthijskooijman reviewed (commented):
> Sorry for letting this languish for so long, I had other work to attend to.

Don't worry, I know the position all too well (and have plenty of other work to attend to in the meanwhile :-p)

Your changes look good, I left some more inline comments wrt the new config directive.

> This is true of my particular changes, but the board custom menu options are already retained as part of the existing functionality.

Right, I did notice this before, though it seemed a bit erratic before. Did you look into the code how it works exactly? I have the impression that maybe it just saves the values of all options, even those not applicable to the current board, so that when you you switch form e.g. the nano with cpu=atmega168 selected to the uno which has no options, and then back, it has remembered  cpu=atmega168. But then I'm afraid that if you switch to another board that *does* have a cpu option, change the cpu options for that board, and switch back to the nano through the recent boards menu, then it would still have the cpu option of the new board, rather than the old board.

In some cases, this might actually be what you want, e.g. when you switch from one ESP board to another, then enable the debug "board option" (which is more of a core option, really), and then switch back to the first ESP board. OTOH, maybe this should be solved by making core options core options and leaving board options board options (i.e. https://github.com/arduino/arduino-cli/issues/922).

Responding to what you wrote before editing,

> I can't think of a clean way to represent the different permutations in the menu. Some boards have 4 or more custom menus so while it's not totally outlandish to image an entry like:
"Arduino Pro or Pro Mini [ATmega328P (5V 16MHz)]"
You would also expect entries in excess of:
"SparkFun Artemis Module[SparkFun Variable Bootloader(Recommended),921600]"

This would be even more of a problem with cores like the ESP8266, which have even more options. i.e. here's what the IDE shows in the statusbar for such a board:

![image](https://user-images.githubusercontent.com/194491/97013832-d416ec00-1549-11eb-8022-5d5a4220e857.png)

But indeed, I think we agree that including the same board with different options is not feasible, so let's not :-)


Wrt the two optimizations I suggested before, I'll have a look myself to see if I can manage to implement those. If I do, ok if I just add a commit to your branch?

matthijskooijman commented:
> Right, I did notice this before, though it seemed a bit erratic before. Did you look into the code how it works exactly? 

I looked at this, it seems that the core of this code is [here](https://github.com/arduino/Arduino/blob/a056a5fdc751b933453830434e36aeffae7512d5/app/src/processing/app/Base.java#L1558-L1633) and indeed mixes up settings for different boards and/or cores and is a bit of a mess in general. I'm halfway through writing an issue that points out some of the problems with the current approach, that could maybe be solved together with this PR, so I'll come back to this somewhere next week.

NPoole commented:
> Wrt the two optimizations I suggested before, I'll have a look myself to see if I can manage to implement those. If I do, ok if I just add a commit to your branch?

Please feel free!

matthijskooijman commented:
See #10887 for the issues I had with the custom boards handling. It also makes the following suggestion to fix part of those inconsistencies:

> To fix this, I would consider just not preserving options at all between different boards and also not when coming back to an earlier board. IOW, whenever you select a different board, just reset all options to the default values.

> One exception could be that for the recently-used-boards list (as being developed in #10816), one would probably expect to get the same options as the last time this particular board was used.

> To implement this, and at the same time heavily simplify implementation, I would suggest removing the current board, target_package, targe_platform and custom_* preferences, and replace them with a single fqbn preference (as already used by arduino-builder and arduino-cli). The recently-used-boards list can then also just store a list of fqbn's and easily include all board options in there.

How does that sound?

I also noticed that the implementation in this PR currently only stores bare board ids, excluding the package or platform (i.e. it stores `uno`, not `arduino:avr:uno`), meaning that when the same board id is used in different platforms, they will get mixed up. This should be fixed by using the full board name, or as suggested above, the `fqbn` including any board menu options.

Because that issue and this PR have some interaction, it might be useful to fix them together (or at least not merge the recently used boards support in one way, and then change how it works wrt to board menus later).

> Wrt the two optimizations I suggested before, I'll have a look myself to see if I can manage to implement those. If I do, ok if I just add a commit to your branch?

Hm, seems I cannot, I think you unchecked the "Allow contributors to push to my branch" or something like that when creating the pullrequest.

So, I pushed to my own branch instead, feel free to pull from there: https://github.com/matthijskooijman/Arduino/commits/recent-boards

I just added some new commits, I think most of the commits can be squashed later before merging this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino/Arduino?pullRequest=10816) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino/Arduino?pullRequest=10816) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino/Arduino?pullRequest=10816) it.</sub>
